docs/proposals/dynamic-app-parameters.md

```diff
@@ -0,0 +1,181 @@
+
+### Unknown kind or group
+
+When the user references a kind or group that does not exist on the destination cluster (possibly a typo in the configuration, etc), a detailed error message may be displayed to the user, since it is assumed that the user would have access to the resource.
+
+### Referenced resource does not exist
+
+When the referenced resource does not exist, and path is not set, it will set the parameter’s value to false. If a path is given, an error is returned to the user that the referenced resource does not exist.
+
+### Field in referenced resource does not exist
+
+When there is a field reference given as path, but this path cannot be found in the referenced resource, a specific error message (e.g. path `.data.foo` does not exist) may be returned to the user, since it is assumed that the user has access to the resource.
+
+## Security implications
+
+### Preventing reading arbitrary cluster resources
+
+Dynamic parameters must not be allowed to read arbitrary resources on the cluster. Instead, there must be some access control mechanisms.
+
+Access control for retrieving values from clusters will be enforced by restrictions in the `AppProject`, some of which already are in place, and will be using the following principles:
+
+* If access to the resource would be allowed through existing `namespaceResourceWhitelist`, `namespaceResourceBlacklist`, `clusterResourceWhitelist` and `clusterResourceBlacklist` configuration, then a dynamic parameter may access the resource as well.
+* Additionally, new access controls specifically for allowing read-only access to dynamic parameters should be introduced. In addition to referencing accessable resources through `kind`, `group` and (possibly) `namespace` attributes, it should optionally be more refinable by a specific resource `name`. The fields should be called `namespaceReadOnlyAllowlist` and `clusterReadOnlyAllowList` respectively, and not be implicitly bound by the field name to the dynamic parameter feature for future applications. There's no need (yet) for a deny list in this scenario.
+
+### Transparency of dynamic parameters' values
+
+It should be either documented that dynamic parameters should not refer to confidential data stored in a cluster (e.g. a Secret), or we need to take care not to leak any of the values back to a user who has e.g. `get` or `sync` access for the Application. Leakage could occur anywhere, in error messages, logs, etc.
+
+We should keep in mind that even with dynamic parameters, the values will be part of the command line when `helm template` is executed.
+
+Therefore, the recommendation is to make it clear to not use this feature for confidential items.
+
+## Alternatives
+
+None considered so far.
```

Comment:
I think we should consider: 
1) help push [`--dry-run server` support in Helm](https://github.com/helm/helm/pull/9426)
2) switch from `helm template` to `helm install --dry-run server`
3) [implement project-level SA impersonation](https://github.com/argoproj/argo-cd/issues/7689)

Pros:
* no wheel reinvention
* we add impersonation, a highly-requested feature, useful for more than just Helm

Cons:
* impersonation is a big project
* switching to `helm install` might be a big project

Reply:
Agreed, I will add this to the alternatives section. 

Not re-inventing the wheel is definitely a big plus, on the other hand, dynamic application parameters as originally proposed could also be leveraged to parametrize plugins (and possibly Kustomize) in the future.

I do see some more or less severe security implications for the `helm --dry-run server` feature, as in we would have to make the default repository server instance mount a Kubernetes SA token by default, and this SA would require impersonation privileges (which basically equals to cluster-admin, or "sudo" permissions AFAIK).

User impersonation for the application controller (which we definitely _should_ do) can also be implemented independently :)

Reply:
Ah, good point. Could we have the application controller pass over the creds at generate-time (preferably something short-lived)? Questions for another proposal, but I'm curious if you know off-hand. :-)

Reply:
I would love to have this feature, @crenshaw-dev your suggesting we don't offer this new API on the Application CR?

Reply:
I think this feature will help deploy 3rd party helm charts (ie prometheus, aws-loadbalancer, etc) that the parameters need to be dynamic based on data stored on a configmap or secret or another CR (crossplane MR)